below/ethtool/src/reader.rs

```diff
@@ -0,0 +1,232 @@
+use std::os::fd::{OwnedFd, AsRawFd};
+use std::str;
+
+use nix::errno::Errno;
+use nix::libc::ioctl;
+use nix::sys::socket::{socket, AddressFamily, SockFlag, SockType};
+
+use crate::{
+    ETHTOOL_GSSET_INFO,
+    ETHTOOL_GSTRINGS,
+    ETHTOOL_GSTATS,
+    ethtool_stringset_ETH_SS_STATS as ETH_SS_STATS,
+    ETH_GSTRING_LEN,
+};
+use crate::errors::Error;
+
+const ETH_GSTRINGS_LEN: usize = ETH_GSTRING_LEN as usize;
+const ETH_GSTATS_LEN: usize = 8;
+
+/// Maximum size of an interface name
+const IFNAME_MAX_SIZE: usize = 16;
+
+/// MAX_GSTRINGS maximum number of stats entries that ethtool can retrieve
+const MAX_GSTRINGS: usize = 8192;
+
+#[derive(Debug)]
+#[repr(C)]
+struct StringSetInfo {
+    cmd: u32,
+    reserved: u32,
+    mask: u32,
+    data: usize,
+}
+
+#[derive(Debug)]
+#[repr(C)]
+struct GStrings {
+    pub cmd: u32,
+    pub string_set: u32,
+    pub len: u32,
+    pub data: [u8; MAX_GSTRINGS * ETH_GSTRINGS_LEN],
+}
+
+#[derive(Debug)]
+#[repr(C)]
+struct GStats {
+    pub cmd: u32,
+    pub len: u32,
+    pub data: [u8; MAX_GSTRINGS * ETH_GSTATS_LEN],
+}
+
+#[derive(Debug)]
+#[repr(C)]
+struct IfReq {
+    if_name: [u8; IFNAME_MAX_SIZE],
+    if_data: usize,
+}
```

Comment:
All of these should come from the corresponding C header, no?

Reply:
I don't think a corresponding struct is available in `ethtool.h`